aggregate6.7
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "AGGREGATE6" "7" "December 2017" "" ""
aggregate6 will compress an unsorted list of IP prefixes (both IPv4 and IPv6)\.
.
.SH "Description"
Takes a list of IPv6 prefixes in conventional format on stdin, and performs two optimisations to attempt to reduce the length of the prefix list\. The first optimisation is to remove any supplied prefixes which are superfluous because they are already included in another supplied prefix\. For example, \fB2001:67c:208c:10::/64\fR would be removed if \fB2001:67c:208c::/48\fR was also supplied\.
.
.P
The second optimisation identifies adjacent prefixes that can be combined under a single, shorter\-length prefix\. For example, \fB2001:67c:208c::/48\fR and \fB2001:67c:208d::/48\fR can be combined into the single prefix \fB2001:67c:208c::/47\fR\.
.
.P
The above optimalisation steps are often useful in context of compressing firewall rules or BGP prefix\-list filters\.
.
.P
The following command line options are available:
.
.IP "" 4
.
.nf

    \-4          Only output IPv4 prefixes
    \-6          Only output IPv6 prefixes
    \-h, \-\-help  show help message and exit
    \-m N        Sets the maximum prefix length for entries read, longer prefixes will be discarded prior to processing
    \-t          truncate IP/mask to network/mask
    \-v          Display verbose information about the optimisations
    \-V          Display aggregate6 version
.
.fi
.
.IP "" 0
.
.SH "Installation"
OpenBSD 6\.3:
.
.P
\fB$ doas pkg_add aggregate6\fR
.
.P
Other platforms:
.
.P
\fB$ pip3 install aggregate6\fR
.
.SH "CLI Usage"
Either provide the list of IPv4 and IPv6 prefixes on STDIN, or give filenames containing lists of IPv4 and IPv6 prefixes as arguments\.
.
.IP "" 4
.
.nf

$ # via STDIN
$ cat file_with_list_of_prefixes | aggregate6
   \.\.\. output \.\.\.

$ # with a filename as argument
$ aggregate6 file_with_list_of_prefixes [ \.\.\. optional_other_prefix_lists ]
   \.\.\. output \.\.\.

$ # Whitespace separated works too
$ echo 2001:67c:208c::/48 2000::/3 | aggregate6
2000::/3

$ # You can combine IPv4 and IPv6
$ echo 10\.0\.0\.0/16 10\.0\.0\.0/24 2000::/3 | aggregate6
10\.0\.0\.0/16
2000::/3
.
.fi
.
.IP "" 0
.
.SH "Library Usage"
Aggregate6 can be used in your own pyp/python2/python3 project as python module\. Currently there is just one simple public function: \fBaggregate()\fR which takes a list as parameter\.
.
.IP "" 4
.
.nf

>>> import from aggregate6 import aggregate
>>> aggregate(["10\.0\.0\.0/8", "10\.0\.0\.0/24"])
[\'10\.0\.0\.0/8\']
>>>
.
.fi
.
.IP "" 0
.
.SH "Bugs"
Please report bugs at https://github\.com/job/aggregate6/issues
.
.SH "Author"
Job Snijders \fIjob@instituut\.net\fR


aggregate6/aggregate6.py
#!/usr/bin/env python
# Copyright (C) 2014-2017 Job Snijders <job@instituut.net>
#
# This file is part of aggregate6
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

from __future__ import print_function
from __future__ import unicode_literals

from builtins import str as text
from ipaddress import ip_network, ip_interface

import aggregate6
import radix
import sys


def aggregate(l):
    """Aggregate a `list` of prefixes.

    Keyword arguments:
    l -- a python list of prefixes

    Example use:
    >>> aggregate(["10.0.0.0/8", "10.0.0.0/24"])
    ['10.0.0.0/8']
    """
    tree = radix.Radix()
    for item in l:
        try:
            tree.add(item)
        except (ValueError) as err:
            raise Exception("ERROR: invalid IP prefix: {}".format(item))

    return aggregate_tree(tree).prefixes()


def aggregate_tree(l_tree):
    """Walk a py-radix tree and aggregate it.

    Arguments
    l_tree -- radix.Radix() object
    """

    def _aggregate_phase1(tree):
        # phase1 removes any supplied prefixes which are superfluous because
        # they are already included in another supplied prefix. For example,
        # 2001:67c:208c:10::/64 would be removed if 2001:67c:208c::/48 was
        # also supplied.
        n_tree = radix.Radix()
        for prefix in tree.prefixes():
            if tree.search_worst(prefix).prefix == prefix:
                n_tree.add(prefix)
        return n_tree

    def _aggregate_phase2(tree):
        # phase2 identifies adjacent prefixes that can be combined under a
        # single, shorter-length prefix. For example, 2001:67c:208c::/48 and
        # 2001:67c:208d::/48 can be combined into the single prefix
        # 2001:67c:208c::/47.
        n_tree = radix.Radix()
        for rnode in tree:
            p = text(ip_network(text(rnode.prefix)).supernet())
            r = tree.search_covered(p)
            if len(r) == 2:
                if r[0].prefixlen == r[1].prefixlen == rnode.prefixlen:
                    n_tree.add(p)
                else:
                    n_tree.add(rnode.prefix)
            else:
                n_tree.add(rnode.prefix)
        return n_tree

    l_tree = _aggregate_phase1(l_tree)

    if len(l_tree.prefixes()) == 1:
        return l_tree

    while True:
        r_tree = _aggregate_phase2(l_tree)
        if l_tree.prefixes() == r_tree.prefixes():
            break
        else:
            l_tree = r_tree
            del r_tree

    return l_tree


def parse_args(args):
    import argparse
    p = argparse.ArgumentParser(description="Aggregate lists of IP prefixes",
                                epilog="""
Copyright 2014-2017 Job Snijders <job@instituut.net>
Project website: https://github.com/job/aggregate6
""", formatter_class=argparse.RawTextHelpFormatter)
    p.add_argument('-v', dest='verbose', action='store_true',
                   help="Display verbose information about the optimisations")
    p.add_argument('-t', dest='truncate', action='store_true',
                   help="truncate IP/mask to network/mask")
    p.add_argument('-m', dest='max_length', metavar='N', type=int,
                   default=0, help="Sets the maximum prefix length for \
entries read, longer prefixes will be discarded prior to processing")
    p.add_argument('-V', dest='version', action='store_true',
                   help="Display aggregate6 version")
    afi_group = p.add_mutually_exclusive_group()
    afi_group.add_argument('-4', dest='ipv4_only', action='store_true',
                           default=False,
                           help="Only output IPv4 prefixes")
    afi_group.add_argument('-6', dest='ipv6_only', action='store_true',
                           default=False,
                           help="Only output IPv6 prefixes")
    p.add_argument('args', nargs=argparse.REMAINDER,
                   help='[file ...] or STDIN')
    return p.parse_args(args)


def main():
    import fileinput

    args = parse_args(sys.argv[1:])

    if args.version: # pragma: no cover
        print("aggregate6 %s" % aggregate6.__version__)
        sys.exit()

    p_tree = radix.Radix()

    for line in fileinput.input(args.args):
        if not line.strip(): # pragma: no cover
            continue
        for elem in line.strip().split():
            try:
                if args.truncate:
                    prefix_obj = ip_interface(text(elem.strip())).network
                else:
                    prefix_obj = ip_network(text(elem.strip()))
                prefix = text(prefix_obj)
            except (ValueError) as err:
                sys.stderr.write("ERROR: '%s' is not a valid IP network, \
ignoring.\n" % elem.strip())
                continue

            if args.max_length > 0:
                if prefix_obj.prefixlen > args.max_length:
                    continue

            if args.ipv4_only and prefix_obj.version == 4:
                p_tree.add(prefix)
            elif args.ipv6_only and prefix_obj.version == 6:
                p_tree.add(prefix)
            elif not args.ipv4_only and not args.ipv6_only:
                p_tree.add(prefix)

    if args.verbose:
        input_list = p_tree.prefixes()
        output_list = aggregate_tree(p_tree).prefixes()
        for p in sorted(set(input_list + output_list)):
            if p in input_list and p not in output_list:
                print("- ", end='')
            elif p in output_list and p not in input_list:
                print("+ ", end='')
            else:
                print("  ", end='')
            print(p)
    else:
        for prefix in aggregate_tree(p_tree).prefixes():
            print(prefix)
